### PR TITLE
feat(notifications/pubsub): Adds support for Google Pubsub notifications

### PR DESCRIPTION
**Pull request by ttomsu**
Sorry this PR is huge - there were a lot of points to hook into the infrastructure and I didn't realize how big it was getting.

tag: https://github.com/spinnaker/spinnaker/issues/2784

### CODE REVIEW

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractAddPublisherCommand.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.netflix.spinnaker.halyard.cli.command.v1.config.notifications.pubsub;
```
Comment (ezimanyi) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractAddPublisherCommand.java]: Missing copyright header

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractDeletePublisherCommand.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017 Google, Inc.
```
Comment (ezimanyi) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractDeletePublisherCommand.java]: Wrong year

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2017 Google, Inc.
```
Comment (ezimanyi) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java]: Wrong year

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractAddPublisherCommand.java (hunk 1 above)
Comment (ttomsu) [reply halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractAddPublisherCommand.java]: Done here and all missing files

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractDeletePublisherCommand.java (hunk 2 above)
Comment (ttomsu) [reply halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractDeletePublisherCommand.java]: Done.

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java (hunk 3 above)
Comment (ttomsu) [reply halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java]: Done.

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    String pubsubName = getPubsubName();
+    String currentDeployment = getCurrentDeployment();
+    // Disable validation here, since we don't want an illegal config to prevent us from fixing it.
+    Publisher publisher = new OperationHandler<Publisher>()
+        .setFailureMesssage(
+            "Failed to get publisher " + publisherName + " for pubsub " + pubsubName + ".")
+        .setOperation(Daemon.getPublisher(currentDeployment, pubsubName, publisherName, false))
+        .get();
+
+    int originaHash = publisher.hashCode();
```
Comment (joonlim) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/command/v1/config/notifications/pubsub/AbstractEditPublisherCommand.java]: originalHash

File: halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/services/v1/DaemonService.java [hunk 5]
```diff
@@ -254,6 +254,40 @@
...
+  DaemonTask<Halconfig, Void> deletePublisher(
+      @Path("deploymentName") String deploymentName,
+      @Path("pubsubName") String pubsubName,
+      @Path("publisherName") String publisherName,
+      @Query("validate") boolean validate);
+
+
+
+
+
```
Comment (joonlim) [inline halyard-cli/src/main/java/com/netflix/spinnaker/halyard/cli/services/v1/DaemonService.java]: Please remove newlines

**Pull request merged**